src/odemis/util/comp.py

```diff
@@ -113,64 +115,67 @@ def get_fov_rect(comp, fov):
-        returned list is adjusted to the closest value of the given zstep that divides
-        the zrange. If zstep is too small, this will lead to too large number of zlevels. 
-        In that case IndexError is raised. 
-    returns (list of floats): list of zlevels, where each zlevel is absolute position for
-        the focuser.  
+    and zstep, as well as the current focus z position.
+    :param focuser: Actuator component with a "z" axis, to control the focus.
+    :param zrange: contains the zmin and zmax, respectively. It's relative to the
+    current position of the focuser. If the range would go out of the actuator
+    range, it's clipped (so there are less zlevels returned than expected).
+    Error is raised if the order is not (zmin, zmax).
+    :param zstep: distance between two successive zlevels. If negative,
+    the order of the returned list of zlevels is reversed. The actual zstep in the
+    returned list is adjusted to the closest value of the given zstep that divides
+    the zrange. If zstep is too small, this will lead to too large number of zlevels.
+    In that case IndexError is raised.
+    :returns: list of zlevels, where each zlevel is absolute position for
+    the focuser.
     """
     if zstep == 0:
-        raise ZeroDivisionError("The step size 'Zstep' can not be zero")
+        raise ZeroDivisionError("The step size 'zstep' can not be zero")
     if zrange[0] > zrange[1]:
-        raise ValueError("The given range is not correct. zmin should be first and zmax second")
+        raise ValueError(f"The given range {zrange} is not correct. The first value should be smaller than the second one.")
     if "z" not in focuser.axes.keys():
-        raise KeyError("The focus actuator %s does not have z axis" %focuser)
+        raise KeyError(f"The focus actuator {focuser} does not have z axis")
 
     focuser_pos = focuser.position.value["z"]
+    zrange_abs = (zrange[0] + focuser_pos, zrange[1] + focuser_pos)
 
     # Get the range from the axis range + extra limit POS_ACTIVE_RANGE
     focuser_rng = focuser.axes["z"].range
     sw_rng = focuser.getMetadata().get(model.MD_POS_ACTIVE_RANGE)
-    if sw_rng:
-        focuser_rng = (max(focuser_rng[0], sw_rng[0]),
-                       min(focuser_rng[1], sw_rng[1]))
-
-    # clip the zMax and zMin to the actuator limits if necessary 
-    if zrange[1] + focuser_pos > focuser_rng[1]:
-        zrange[1] = focuser_rng[1] - focuser_pos
-    if focuser_pos + zrange[0] < focuser_rng[0]:
-        zrange[0] = focuser_rng[0] - focuser_pos
-    if zrange[0] == zrange[1]:
+    if sw_rng and "z" in sw_rng:
+        focuser_rng = (max(focuser_rng[0], sw_rng["z"][0]),
+                       min(focuser_rng[1], sw_rng["z"][1]))
+
+    # clip the zMax and zMin to the actuator limits if necessary
+    zrange_abs = (min(max(focuser_rng[0], zrange_abs[0]), focuser_rng[1]),
+                  min(max(focuser_rng[0], zrange_abs[1]), focuser_rng[1]))
```

Comment:
This always breaks my head a bit, but shouldn't it be:
```suggestion
    zrange_abs = (min(max(focuser_rng[0], zrange_abs[0]), focuser_rng[1]),
                  max(min(focuser_rng[1], zrange_abs[1]), focuser_rng[0]))
```

Reply:
No. Here, the two lines do clipping of the zrange value between the extrema of focuser_rng. So both lines should be identical, except of the zrange index.
Clipping is done as `min(max(low, x), high)`.
Maybe instead of open-coding it every time, we should add it to odemis.util.clip (we already have a version for the VAs, and numpy has its own version, but we don't have a generic one).